packages/@lwc/template-compiler/src/parser/parser.ts

```diff
@@ -53,13 +49,18 @@ interface ParentWrapper {
 }
 
 interface IfContext {
-    ifBlockParent: IfBlock;
-    seenSlots: Set<string>;
+    currentNode: IfBlock | ElseifBlock | ElseBlock;
+
+    // Within a specific if-context, each set of seen slot names must be tracked separately
+    // because duplicate names in separate branches of the same if-block are allowed (the branching
+    // logic provides a compile-time guarantee that the slots will not be rendered multiple times).
+    // seenSlots keeps track of each set holding the slots seen in each branch of the if-block.
+    seenSlots: Set<string>[];
 }
 
-interface SiblingContext {
-    nodes: ParentNode[];
+interface SiblingScope {
     ifContext?: IfContext;
+    ancestorIfContext?: IfContext;
 }
```

Comment:
For a future PR, I think we can combine these two contexts with the `elementScopes` and `siblingContext` to create a broader `scope` object to store everything.

Reply:
Agreed, I think there's definitely room for a few more iterations of refactoring.